sdk/core/src/pipeline_context.rs

```diff
@@ -9,19 +9,19 @@ use crate::Context;
 /// long as it's Send and Sync. For example, Cosmos uses the
 /// PipelineContext to pass the ResourceType down to its
 /// AuthorizationPolicy.
-pub struct PipelineContext<C>
+pub struct PipelineContext<'a, C>
 where
     C: Send + Sync,
 {
-    inner_context: Context,
+    inner_context: &'a mut Context,
```

Comment:
Adding this complicates the code *a lot*. Is there a reason we need to change this to a mutable reference as opposed to an owned value?

Reply:
Unfortunately yes. We need have a way for the pipeline to influence the context. 
In the case of the mock framework, the pipeline increments the "operation number" counter at each Azure interaction. The number is in turn used to "order" the operations in a transaction. As it is now, the context is "lost" because we are handing the pipeline a fresh context each time (and the pipeline doesn't give it back). If we don't give back the updated context, each call will start the transaction at zero. 

The cool thing is, the SDK users just need to keep passing the context to every call and the mock framework will take care of everything. They do not need to concern themselves of the *contents* of the `Context` the just need to pass it around. 

In the previous repo I used to store that information in environment variables but we decided against it. The added benefit of this new `&mut Context` approach is that the framework is now fully multithread capable.